Remove btrfs quota groups after containers destroyed

yongtang opened the pull request:
**- What I did**

This fix tries to address the issue raised in #29325 where btrfs quota groups are not clean up even after containers have been destroyed.

**- How I did it**

The reason for the issue is that btrfs quota groups have to be explicitly destroyed. This fix fixes this issue.

**- How to verify it**

This fix is tested manually in Ubuntu 16.04:

**pre-fix behavior with qgroup leftover:**

```
ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
b59e6e3da6ed3e601f081c11c44d3f4dd44aefee5584d34ba913e68207c9694d
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         144.00KiB    144.00KiB
0/257         1.17MiB      1.17MiB
0/258           0.00B        0.00B
0/259           0.00B        0.00B

```

**PR with qgroup cleaned (non --live-restore case):**
```
ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
201d444f74943247c7794fae2215f763cad0dd726d03285004c54ffa04882d6b
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         144.00KiB    144.00KiB
0/257         1.17MiB      1.17MiB
```


**PR with qgroup cleaned (--live-restore case):**
```
# [sudo dockerd --debug --live-restore -s btrfs --storage-opt btrfs.min_space=1G]

ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
73c2184a4414d47b5f5dc27cd2ee28116aa761f947196618ca66bf05c4705456
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$

# [Ctrl-C ]
# [sudo dockerd --debug --live-restore -s btrfs]

ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS              PORTS               NAMES
73c2184a4414        busybox             "top"               About a minute ago   Up About a minute                       top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         152.00KiB    152.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     48.00KiB
0/259         1.17MiB     48.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         168.00KiB    168.00KiB
0/257         1.17MiB     64.00KiB
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![c03904d6028381a6bd0d1b1bfd1523bb](https://cloud.githubusercontent.com/assets/6932348/21203151/1321041c-c206-11e6-95f8-b237b4bfe945.jpg)


This fix fixes #29325.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Remove btrfs quota groups after containers destroyed

This fix tries to address the issue raised in 29325 where
btrfs quota groups are not clean up even after containers
have been destroyed.

The reason for the issue is that btrfs quota groups have
to be explicitly destroyed. This fix fixes this issue.

This fix is tested manually in Ubuntu 16.04,
with steps specified in 29325.

This fix fixes 29325.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
/cc @justincormack @mlaventure

zhuguihua commented:
Sorry for late reply. LGTM.

yongtang commented:
Previously the PR didn't take into consideration that, if `quotaEnabled` is false, there is no need to clean up the qgroup (See #29810 for similar issue)

Now the PR has been updated to add the check for  `quotaEnabled`.

thaJeztah commented:
Also wondering the same here as on https://github.com/docker/docker/pull/29835#discussion_r94377506

mlaventure commented:
I had a quick look at the code, it looks like @thaJeztah may be correct.

Would wait on @cpuguy83 for confirmation

thaJeztah commented:
_if_ I'm correct, we may need some mechanism to detect if a quota is set (perhaps querying btrfs), otherwise it's possible we don't know when to cleanup / manage quotas

cpuguy83 commented:
I think you are right. On daemon restore it could really be a problem.

LK4D4 commented:
@cpuguy83 @tonistiigi @yongtang so, what we're going to do with this and #29385
If we have no proper solution for now, I propose to close PRs.
Thanks!

yongtang commented:
@cpuguy83 @tonistiigi @LK4D4 This PR (#29427) and the other PR (#29835) are related.

I updated the other PR #29835 first to try to address the issue by looking up the qgroup and see if it returns something.

Please take a look at the (#29835) first and see if it address the concerns.

yongtang commented:
Here is the complete steps to verify that the issue is fixed (Including --live-restore case). 

*Would be good to setup integration test for btrfs in Jenkins*

*Note: this PR has been updated to incorporate #30497. If this one is merged then #30497 could be closed*


**pre-fix behavior with qgroup leftover:**

```
ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
b59e6e3da6ed3e601f081c11c44d3f4dd44aefee5584d34ba913e68207c9694d
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         144.00KiB    144.00KiB
0/257         1.17MiB      1.17MiB
0/258           0.00B        0.00B
0/259           0.00B        0.00B

```

**PR with qgroup cleaned (non --live-restore case):**
```
ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
201d444f74943247c7794fae2215f763cad0dd726d03285004c54ffa04882d6b
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         144.00KiB    144.00KiB
0/257         1.17MiB      1.17MiB
```


**PR with qgroup cleaned (--live-restore case):**
```
# [sudo dockerd --debug --live-restore -s btrfs --storage-opt btrfs.min_space=1G]

ubuntu@ubuntu:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
4b0bc1c4050b: Pull complete
Digest: sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e
Status: Downloaded newer image for busybox:latest
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257       112.00KiB    112.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker run --name top -d busybox top
73c2184a4414d47b5f5dc27cd2ee28116aa761f947196618ca66bf05c4705456
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         128.00KiB    128.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     16.00KiB
0/259         1.17MiB     16.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$

# [Ctrl-C ]
# [sudo dockerd --debug --live-restore -s btrfs]

ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS              PORTS               NAMES
73c2184a4414        busybox             "top"               About a minute ago   Up About a minute                       top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         152.00KiB    152.00KiB
0/257         1.17MiB     64.00KiB
0/258         1.17MiB     48.00KiB
0/259         1.17MiB     48.00KiB
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker rm -f top
top
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$
ubuntu@ubuntu:~$ sudo btrfs qgroup show /var/lib/docker/btrfs
qgroupid         rfer         excl
--------         ----         ----
0/5         168.00KiB    168.00KiB
0/257         1.17MiB     64.00KiB
```

mlaventure commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -320,7 +337,7 @@ func (d *Driver) subvolEnableQuota() error {
 func (d *Driver) subvolDisableQuota() error {
 	if !d.quotaEnabled {
```
Shouldn't this be turned into a `isQuotaEnabled()` routine to avoid this code to be duplicated?

Also, `quotaEnabled` is just a boolean, but don't we need a lock around it?

mlaventure commented:
ping @cpuguy83 @tonistiigi

yongtang commented:
Thanks @mlaventure for the review. The PR has been updated with `isQuotaEnabled()` and lock added. Please take a look and let me know if there are any issues.

mlaventure commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -288,13 +307,22 @@ func subvolDelete(dirpath, name string) error {
+}
+
+func (d *Driver) subvolEnableQuota() error {
+	d.mu.Lock()
```
move the lock in `isQuotaEnabled`?

yongtang replied:
Thanks @mlaventure.

Inside the `subvolEnableQuota` the `d.quotaEnabled` will be changed again to `true` at the end of `subvolEnableQuota` (located in Line 343). For that I placed the `Lock()` outside of `isQuotaEnabled()`.

mlaventure replied:
Oh, my bad. Thanks for clarifying.

mlaventure commented:
ping @cpuguy83 @tonistiigi this is waiting for reviews  :angel:

thaJeztah commented:
ping @cpuguy83 @tonistiigi PTAL

tonistiigi commented:
I think implementation could be cleaner with `sync.Once` but LGTM

yongtang commented:
Thanks @tonistiigi for the review. The PR has been updated to use `sync.Once`. Please take a look.

tonistiigi commented:
@yongtang Sorry for the delay. If the `updateQuotaStatus` is always called inside `once` you should be able to just move the `once` inside it.
Looking at the quota setting, it doesn't seem to be quite safe in case of parallel `Create` calls, but that isn't caused by this PR.

yongtang commented:
Thanks @tonistiigi. The PR has been updated.

tonistiigi commented:
LGTM

vdemeester reviewed (approved):
LGTM 🐯 
/cc @cpuguy83 @mlaventure

sargun commented:
Does anyone see risk around applying this back to 17.04?

thaJeztah commented:
@sargun its too late for docker 17.04, so will end up in 17.05, which will be released a month from now. Of course, it's possible to build it yourself

cpuguy83 commented:
This is not re-applying qutoas on daemon restart.


```
$ docker run -d --name test --storage-opt size=100m busybox top
$ docker exec test sh -c 'dd if=/dev/urandom of=/data bs=1 count=200' # errors around 100M
$ # restart docker
$ # note that `btrfs qgroup show` does not find the qgroup
$ docker start test
$ # note still no qgroup
$ docker exec test sh -c 'dd if=/dev/urandom of=/data bs=1 count=200'  # write successful, no qutoa
```

cpuguy83 commented:
It looks like we need to rescan for quota in `Get()`... or not disable quota on `Cleanup()`

cpuguy83 commented:
Or I guess write out a file to trigger an auto-enable of the quota on restart.

mlaventure commented:
ping @yongtang :angel:

thaJeztah commented:
ping @yongtang do you still have time to work on this?

yongtang commented:
@thaJeztah Sorry for the late response. Let me spend the next several days to get this one fixed.

yongtang committed:
Persist the quota size for btrfs so that daemon restart keeps quota

This commit is an extension of fix for 29325 based on the review comment.
In this commit, the quota size for btrfs is kept in `/var/lib/docker/btrfs/quotas`
so that a daemon restart keeps quota.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@cpuguy83 The PR has been updated. Now the quota is saved in `/lib/var/docker/btrfs/quotas` and recovered in case of a daemon restart. Not sure this is the best place to persist so any suggestions would be greatly appreciated. Please take a look.

thaJeztah commented:
ping @cpuguy83 

also @cyphar, who may be interested in this 😃

sargun commented:
Alternatively, you can just wait for this patchset to land: http://www.spinics.net/lists/linux-btrfs/msg65928.html

thaJeztah commented:
@sargun nice! Unfortunately we don't control the host, so cannot guarantee that patch is present on each distro / version, so not sure we can rely on that

cpuguy83 commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -557,6 +635,13 @@ func (d *Driver) Get(id, mountLabel string) (string, error) {
+	if quota, err := ioutil.ReadFile(d.quotasDirID(id)); err == nil {
+		if size, err := strconv.ParseUint(string(quota), 10, 64); err == nil && size >= d.options.minSpace {
+			d.subvolEnableQuota()
+			subvolLimitQgroup(dir, size)
```
error check?

yongtang replied:
Done.

cpuguy83 commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -557,6 +635,13 @@ func (d *Driver) Get(id, mountLabel string) (string, error) {
 
+	if quota, err := ioutil.ReadFile(d.quotasDirID(id)); err == nil {
+		if size, err := strconv.ParseUint(string(quota), 10, 64); err == nil && size >= d.options.minSpace {
+			d.subvolEnableQuota()
```
error check?

yongtang replied:
Done.

cpuguy83 commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -533,7 +606,12 @@ func (d *Driver) Remove(id string) error {
 		return err
 	}
-	if err := subvolDelete(d.subvolumesDir(), id); err != nil {
+	os.Remove(d.quotasDirID(id))
```
error check?

yongtang replied:
Thanks @cpuguy83. Done.

cpuguy83 commented on daemon/graphdriver/btrfs/btrfs.go:
```diff
@@ -466,9 +532,16 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) error {
+		if err := idtools.MkdirAllAs(quotas, 0700, rootUID, rootGID); err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(path.Join(quotas, id), []byte(fmt.Sprint(driver.options.size)), 0666); err != nil {
```
644?

yongtang replied:
Thanks @cpuguy83. The PR has been updated.

yongtang commented:
Thanks @cpuguy83 for the review. The PR has been updated. Please take a look.

cpuguy83 reviewed (approved):
LGTM

yongtang commented:
All tests passed now, merging...
